opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -551,13 +553,16 @@ def get_span_context(self):
 
     def set_attribute(self, key: str, value: types.AttributeValue) -> None:
         with self._lock:
-            if not self.is_recording():
-                return
+            is_recording = self.is_recording()
```

Comment:
Why change this logic?

Reply:
Intent is to preserve the existing behavior of emitting a logger warning if you call the method on an ended span. Since is_recording is false on an ended span, it required changing the order of the checks - otherwise we'd never hit the has_ended check.

I went back and forth on this since it changes the convention of checking is_recording first and returning quick from the various methods where this is done, but felt the benefit of maintaining the existing behavior from a caller perspective with receiving the warning was worth it.